libc/src/stdlib/block.h

```diff
@@ -0,0 +1,482 @@
+//===-- Implementation header for a block of memory -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC_STDLIB_BLOCK_H
+#define LLVM_LIBC_SRC_STDLIB_BLOCK_H
+
+#include "src/__support/CPP/algorithm.h"
+#include "src/__support/CPP/cstddef.h"
+#include "src/__support/CPP/limits.h"
+#include "src/__support/CPP/new.h"
+#include "src/__support/CPP/optional.h"
+#include "src/__support/CPP/span.h"
+#include "src/__support/CPP/type_traits.h"
+
+#include <stdint.h>
+
+namespace LIBC_NAMESPACE {
+
+namespace internal {
+// Types of corrupted blocks, and functions to crash with an error message
+// corresponding to each type.
+enum class BlockStatus {
+  kValid,
+  kMisaligned,
+  kPrevMismatched,
+  kNextMismatched,
+};
+} // namespace internal
+
+/// Returns the value rounded down to the nearest multiple of alignment.
+LIBC_INLINE constexpr size_t AlignDown(size_t value, size_t alignment) {
+  __builtin_mul_overflow(value / alignment, alignment, &value);
+  return value;
+}
+
+/// Returns the value rounded down to the nearest multiple of alignment.
+LIBC_INLINE template <typename T>
+constexpr T *AlignDown(T *value, size_t alignment) {
+  return reinterpret_cast<T *>(
+      AlignDown(reinterpret_cast<size_t>(value), alignment));
+}
+
+/// Returns the value rounded up to the nearest multiple of alignment.
+LIBC_INLINE constexpr size_t AlignUp(size_t value, size_t alignment) {
+  __builtin_add_overflow(value, alignment - 1, &value);
```

Comment:
if the overflow bit is not used, you can just use the normal `value + (alignment - 1)`

Reply:
another option to avoid overflow is:
```
  if (value == 0)
    return 0;
  return ((value - 1) / alignment + 1) * alignment;
```

Reply:
Does llvm-libc have something equivalent to a `DEBUG_ASSERT` macro where if we're in "debug mode", we do a normal assertion but if in "release mode", we still evaluate the thing in the macro but ignore the result? Prior, this line looked something like `DEBUG_ASSERT(__builtin_add_overflow(value, alignment - 1, &value))` so we expect it to never overflow,  but just in case it would be nice to check. If not, I can just use `value + (alignment - 1)`.

Reply:
so I think if `value > floor(SIZE_MAX / alignment)` then overflow always happens.  The question is whether do you want to be safe or ignore that case?

Reply:
I think we can ignore that case for here since we likely won't ever end up with a size that high.